### ISSUE TITLE
Improve pytest completions and goto def

### ISSUE THREAD

**Issue by rchiodo** [microsoft/pylance-release#3727]
Mini-spec:

- Go to def on a test parameter finds the fixture
- Fixture types are known (and help with writing a test)
- All known fixtures are listed in the completions for parameters to a function (if inside a test function)
- Parameters of a test that aren't found in the list of fixtures should show an error/warning

Example - Hover over a parameter on a test function or a fixture shows the type of the parameter:

![pytest1](https://user-images.githubusercontent.com/19672699/208200337-7e906c8d-f41b-4bf1-9d0a-3c98050a70a3.gif)

Example - Go to def works for parameters for tests and fixtures:

![pytest2](https://user-images.githubusercontent.com/19672699/208200504-a3911c6a-f002-4457-9c5c-1a38dfcf1754.gif)

Example - completions are listed in the parameters for a test function or a fixture

![pytest3](https://user-images.githubusercontent.com/19672699/208200643-4ac87b3e-34cb-4801-bec7-3768b9a8cc99.gif)

Example - error for parameters that don't match an existing fixture:
![pytest4](https://user-images.githubusercontent.com/19672699/208200803-b30c2786-fb40-4ac2-9ac2-d3e71bc2e09a.gif)

**Comment by rchiodo** [comment]
Example:
https://github.com/microsoft/pyright/issues/1702

**Comment by rchiodo** [comment]
Pytest raises exceptions. Message regex might be computable?

https://docs.pytest.org/en/7.2.x/reference/reference.html#pytest.raises

**Comment by rchiodo** [comment]
Problems to solve:

- Go to def on a test parameter finds the fixture
- Fixture types are known (and help with writing a test)
- All known fixtures are listed in the completions for parameters to a function

Solution thoughts:

- Pytest separate lerna package
- TypeEvaluator has a hook for param type.
   - Hook needs typeevaluator to get type of fixtures though. It just needs to find the location of the fixture in the parsetree
- Goto def has a hook
- Completion provider has a hook

**Comment by rchiodo** [comment]
[Markers](https://docs.pytest.org/en/7.2.x/reference/reference.html#marks-ref) allow for:

- Labeling tests (not quite so useful for type checking)
- [use](https://docs.pytest.org/en/7.2.x/how-to/fixtures.html#use-fixtures-in-classes-and-modules-with-usefixtures) fixtures with a marker instead of an argument (might want to autocomplete these)
- [parmeterize](https://docs.pytest.org/en/7.2.x/how-to/parametrize.html#parametrize) a function. Parameters are passed as lists of tuples. Might typecheck them? Seems rather difficult.
- skip,skipif,xfail - not useful for us
- [custom](https://docs.pytest.org/en/7.2.x/reference/reference.html#custom-marks) which allow generating more values to be sent into the test

**Comment by rchiodo** [comment]
Switch algorithm to look for test file names first
Check pytest config in Visual Studio

**Comment by rchiodo** [comment]
TypeEval class changes
- Performance tests in Power BI 
- Run pyright CLI against
   - numpy
   - pandas
   - scikit-learn

**Comment by rchiodo** [comment]
Npm Run tests results:

- TypeEvaluator as a class: 196s, 164s, 164s
- TypeEvaluator as a set of functions: 188s, 162s, 163s

**Comment by rchiodo** [comment]
Alternative ideas without changing the TypeEvaluator:

**Idea 1: Second language server**
- Second language server that provides: 
   -  Goto def for pytest fixtures
   -  Hover for pytest fixtures (computes type on its own)
   -  Completions for pytest functions (list fixtures)

Pros:
- Wouldn't violate principles

Cons:
- Probably be a copy of pyright just to do all the parsing and types. Fork? THat would be ugly to maintain.
   - Maybe could add the extensibility just for this separate language server
- Hover would have two types - Any from pylance, and fixture type from new server

**Idea 2: Pytest extension that injects middleware in python core extension?**
- Overrides necessary functionality to customize it for pytest?
- Still needs to compute type somehow, but maybe could use pyright and ask it for the appropriate type. Runs pyright as a server and uses it to answer questions about different parts.

Pros:
- Wouldn't violate principles. 
- Users would install it only for pytest

Cons:
- Separate copy of all of the types
- Maybe instead send messages directly to pylance to compute the types etc. Middleware would only need to do parsing

**Comment by rchiodo** [comment]
I think the pytest weirdness can be summed up as:

- Parameter types come from function return values that match the parameter name.
- These types are applied to only certain functions (having a decorator or matching a pattern)

If you were going to codify that it such a way that a static type checker could figure it out, how would you do it? 


Maybe something like a 'implicit_parameter_types' decorator?


def implicit_parameter_types(
   function_has_name_pattern: str)
   
def implicit_parameter_types(
   function_has_decorator_type: Type)
   
   
Then the hard part would be how to apply this to pytest. For fixtures, we could probably just have some mapping between pytest.fixture and one or more of these, but for 'test_' functions, that doesn't work. Users don't do anything except name functions 'test_'

Not sure how to apply some rule that a type checker can follow to the 'test_' functions. Is there any precedence for general pattern matching and implying something from it?

**Comment by erictraut** [comment]
Pytest fixtures are typically not explicitly imported by the test modules that consume them. (This is a terrible design, IMO. I find pytest tests to be incredibly difficult to understand because of this. It's almost impossible to look at the code and determine what's going on in the test.) Because fixtures are not explicitly referenced in the modules in which they're used, the normal dependency-tracking mechanisms in pyright do not apply. What are your thoughts on how to handle this aspect of the problem? If someone modifies a fixture, how will that change be propagated?

Are you assuming that each project will have only one pytest configuration? What about mono-repos where there are multiple configurations? Have you thought about how to map specific pytest configurations and fixture sets to specific subtrees in the code base?

**Comment by rchiodo** [comment]
Maybe I missed something, but I thought the pytest fixtures are always either:
- In the test file itself
- In a 'conftest.py' file. 

The 'conftest.py' file can be anywhere in the workspace AFAIK, so the plan is to just look up the directory tree to find all the conftest.py files. They're added to the 'tracked' file list in the server.

I need to reread the docs, but maybe there's other ways to define fixtures too.

**Comment by rchiodo** [comment]
Oh yeah, there's this too. [Plugins](https://docs.pytest.org/en/6.2.x/fixture.html#fixtures-from-third-party-plugins)

**Comment by erictraut** [comment]
Ah, so fixtures are discovered based on searching for a hard-coded file name (`conftest.py`) within the current directory structure? I was under the impression that fixture modules (whatever name they were given) were referenced by a pytest configuration file of some sort. My mental model was probably wrong here.

**Comment by rchiodo** [comment]
> Ah, so fixtures are discovered based on searching for a hard-coded file name (`conftest.py`) within the current directory structure? I was under the impression that fixture modules (whatever name they were given) were referenced by a pytest configuration file of some sort. My mental model was probably wrong here.

AFAIK, yes. There might be a way with some sort of outside configuration that I don't know about because @bschnurr mentioned something special that Visual Studio does for pytest. Haven't looked into it yet.

**Comment by erictraut** [comment]
You mentioned that there's a way to invoke a test with a fixture without including a decorator — that pattern matching is also supported. Where is that pattern defined? Is it hard-coded, or is it configurable in some manner?

**Comment by rchiodo** [comment]
Sorry I meant this:

```python
def test_foo(request):
    pass
```

The fixture has no decorator here because it's a builtin one. `request` is not referenced anywhere in the user's source other than as a parameter name.

Conftest.py fixtures are similar too:

```python
# conftest.py
@pytest.fixture
def my_fixture():
   return "4"
```

```python
# test.py
def test_stuff(my_fixture)
   pass
```
`my_fixture` isn't imported or anything. It's just implied.

**Comment by rchiodo** [comment]
Maybe you meant the pattern I mentioned here?

> def implicit_parameter_types(
   function_has_name_pattern: str)
   

Pytest thinks anything starting with the string `test_` is a test function. It will then use the parameter names to match all known fixtures.

**Comment by erictraut** [comment]
Is `test_` hard-coded, or is it configurable? 

Obviously, `test_` has this special meaning only if you're using pytest. Do you have thoughts on how to distinguish between the two cases — one where it should be interpreted as a test with a pytest fixture and one where it doesn't? 

Is it common to rely on `test_` without a decorator, or is this the exception? I'm wondering if the solution can rely on the presence of a `@pytest.fixture` decorator? Could we tell users "you have to use the decorator if you want type support in pylance"? 

I'm trying to get a sense for how much of pytest's internal behaviors need to be assumed here. The problem gets much easier (and the solution less fragile) if we don't need to encode these behaviors.

**Comment by rchiodo** [comment]
> Is test_ hard-coded, or is it configurable?

`test_` is hard coded AFAICT. See [here](https://docs.pytest.org/en/7.2.x/explanation/goodpractices.html#conventions-for-python-test-discovery).

> Obviously, test_ has this special meaning only if you're using pytest. Do you have thoughts on how to distinguish between the two cases — one where it should be interpreted as a test with a pytest fixture and one where it doesn't?

AFAICT, a test with a fixture is any test function that takes a parameter other than self. 

> Is it common to rely on test_ without a decorator, or is this the exception? I'm wondering if the solution can rely on the presence of a @pytest.fixture decorator? Could we tell users "you have to use the decorator if you want type support in pylance"?

Test functions don't have any decorators. [Fixtures](https://docs.pytest.org/en/6.2.x/reference.html#fixtures-api) do. If you want to create a new fixture, you add a decorator. That's the hard part I think in trying to distinguish fixtures as parameters. There's nothing other than the pattern of `test_` on the front and the name of the parameter.

I don't know what we'd tell users other than figure out the types yourself if we don't want pyright to compute them. 

I did have an idea above (Idea 3: Outside command that adds types for features as a quick action) about having a separate action that does the computation for them and then adds the type annotations for them. That way pyright wouldn't have to know about the types, but that seems crummy to me.

**Comment by rchiodo** [comment]
Looks like 'test_' isn't harcoded either. You can have a pytest configuration file too:

https://docs.pytest.org/en/7.2.x/example/pythoncollection.html#changing-naming-conventions

**Comment by erictraut** [comment]
I think we need a more complete functional spec. It looks like you're discovering requirements on the fly, which makes it difficult to understand the scope and bound of the problem that you're trying to solve. Which pytest functionality you think is important to support (and which is not), and how does that translate into pyright/pylance features?

**Comment by rchiodo** [comment]
I don't think we're attempting to fully define everything we want to do or all the scenarios it will cover. The goal is to just get an MVP that we can put in front of customers. 

What do you think is missing?

**Comment by rchiodo** [comment]
Perhaps I should have more examples and what the behavior could be? Like maybe it's hard to imagine what we'd want to have happen for text functions and fixtures.

I'll add some gifs to the top outlining what it could look like.

**Comment by erictraut** [comment]
I'm trying to get a sense for which pytest behaviors need to be understood and replicated in pyright or pylance logic to achieve the desired outcome. Without that understanding, it's difficult to evaluate our technical options.

I'm asking basic questions about how fixtures are applied and evaluated, which behaviors are hard-coded vs configurable, etc., and it sounds like you haven't explored those questions. In my experience, a functional spec is a good way to establish the contours of a problem and make sure that everyone has that shared context.

For example, looking at the pytest test cases in my team's code base, we often use `@pytest.mark.usefixtures` and/or `@pytest.mark.parametrize` decorators. These presumably override which fixtures are applied and would therefore need to be understood by any mechanism that supplies or infers parameter types for such functions.

My understanding of pytest is very incomplete. It sounds like yours is too. We need to obtain a much more complete understanding. Otherwise we risk pursuing a technical approach that falls apart when we discover some new requirement or behavior that needs to be supported.

**Comment by rchiodo** [comment]
A word doc would probably be a better spot to put all of this down. Then I could have stated a bunch of things that I didn't think were going to affect the MVP. 

I didn't think [pytest.mark.usefixtures](https://github.com/microsoft/pylance-release/issues/3727#issuecomment-1341816677) will affect the features proposed here, but now that I think about it more, the `pytest.mark.parameterize` might. 

It'd probably be easier to comment on these things there instead of just having my random thoughts on what _I_ thought the MVP should be. And it could certainly be more exhaustive than what I'm listing here.

**Comment by jwilges** [comment]
Re:
> I believe this is where Graham's idea of generating typestubs might work. Generate a typestub next to the test file with the types filled in based on the fixture return value.

If I'm understanding this suggestion correctly, Graham's suggesting generating *.pyi files and leaning on Pylance picking up completions from the generated [PEP-571](https://peps.python.org/pep-0561/) stub files? That sounds clever at first glance (simple/elegant) but for navigation I don't think anyone would want code navigation to land them in the type stub file. Just an off-the-hip thought while reading over your notes so far.

**Comment by rchiodo** [comment]
> If I'm understanding this suggestion correctly, Graham's suggesting generating *.pyi files and leaning on Pylance picking up completions from the generated [PEP-571](https://peps.python.org/pep-0561/) stub files? That sounds clever at first glance (simple/elegant) but for navigation I don't think anyone would want code navigation to land them in the type stub file. Just an off-the-hip thought while reading over your notes so far.

That was the idea, but the stubs would be in 'virtual file system'. Goto def would likely not use those stubs and would just find the real .py files (as that's what goto def does today for functions that are both in a pyi and a py file).

> If design thoughts steer out of this public thread and into a Word document somewhere, can we follow along somehow?

The word doc (at the moment) is more about internal architecture required to implement the ideas outlined here.

The intent of this issue is to be a first stab at the multiple discussion items, so if the MVP changes, we'll update it here.

**Comment by jwilges** [comment]
> That was the idea, but the stubs would be in 'virtual file system'. Goto def would likely not use those stubs and would just find the real .py files

If that's all viable, it sounds like it could really simplify the MVP here. It may also provide unique avenues to separate the "pytest"-specific concerns from this project, if warranted.

Is this "generate stubs on the fly" approach used anywhere else in "Pylance/Pyright land" so to speak?

Where are the other `pytest`-specific configuration points in VS Code (e.g. `pytest.command`, `python.testing.pytest*`) living today? In Pylance or vscode-python by chance? VS Code at least has a few configuration points today that make `pytest` feel like a bit of a first class citizen in the ecosystem.

(Forgive my ignorance on the above two questions, although I've been enjoying using these projects for a while, I haven't tried to dig into their internals yet, but I figured you might know off hand.)

Pylance principle #2 in your [ideas pros-and-cons list comment](https://github.com/microsoft/pylance-release/issues/3727#issuecomment-1353784369) seems reasonable in that if baking `pytest`-specific behaviors too closely into this project can be avoided, that certainly seems desirable.

The `pytest` runner supports automatically loading plugins (naturally if they are written correctly and [installed](https://docs.pytest.org/en/7.1.x/how-to/plugins.html#how-to-install-and-use-plugins) as well as via setting [`pytest_plugins`](https://docs.pytest.org/en/7.1.x/reference/reference.html#globalvar-pytest_plugins). Perhaps it may be reasonable to have a `pytest` plugin generate fixture stubs in a compatible location/format where VS Code can then shrug off that concern?

**Comment by rchiodo** [comment]
> Is this "generate stubs on the fly" approach used anywhere else in "Pylance/Pyright land" so to speak?

Not at the moment, no. I didn't end up using it for the prototype either as it was more involved than just injecting code into pyright. 

You'd have to generate one for every test function, as the user types them. 

So for example, you have this test func:

```python
def test_foo(my_fixture):
   assert(my_fixture.length == 2)
```

The stub for that might be something like:

```
def test_foo(my_fixture: list[str]): pass
```


> Where are the other pytest-specific configuration points in VS Code (e.g. pytest.command, python.testing.pytest*) living today? In Pylance or vscode-python by chance? VS Code at least has a few configuration points today that make pytest feel like a bit of a first class citizen in the ecosystem. That's what initially made me think this line of features might not be terribly out of place as an integration in Microsoft-official extension(s).

Some are in the vscode-python extension. Some are in [3rd party extensions](https://marketplace.visualstudio.com/search?term=pytest&target=VSCode&category=All%20categories&sortBy=Relevance).

> The pytest runner supports automatically loading plugins via some scanning voodoo I haven't looked up in a while if they are [installed](https://docs.pytest.org/en/7.1.x/how-to/plugins.html#how-to-install-and-use-plugins) or via enabling them explicitly with e.g. [pytest_plugins](https://docs.pytest.org/en/7.1.x/reference/reference.html#globalvar-pytest_plugins). Perhaps it may be reasonable to have a pytest plugin generate fixture stubs in a compatible location/format where VS Code can then shrug off that concern?

Interesting idea that we should keep in mind when designing this. If plugins are used a lot, how do we handle them? I don't know if a plugin would be able to tell us the types of fixtures or not. That's what we'd need in order to generate the type stubs. I don't think the plugin could generate the type stubs because it'd be conflicting with the stubs we're generating.

**Comment by rchiodo** [comment]
Did some more thinking about the type stub generation. There's a bit of a problem with the idea.

Generated type stubs are like a cache of what type information pyright might generate. Pyright does not currently cache anything in between edits. It recomputes everything on every edit. This is because it's famously hard to update the 'right' parts of the cache. So instead it regenerates as much as necessary after every edit. JIT type evaluation is a core tenet of pyright and why it's so damn fast compared to what we had before.

Generating type stubs for test functions would essentially have to be done on every edit. (Same cache hit detection problem). 

That might be way too slow. And it would have to be done before pyright was asked to compute any types (otherwise pyright might go through some point that those stubs affected). So it would have to block all other operations. 

That makes it seem like to me as not the way to do this. 

Granted it might be worth a try just to see how slow it is in practice. All of the types for the predefined fixtures could remain cached (as we can probably assume the user isn't going to update pytest, or we can detect when they do). But types for all user fixtures would have to be transitively computed on every edit.

**Comment by jwilges** [comment]
> The stub for that might be something like:
> 
> ```python
> def test_foo(my_fixture: list[str]): pass
> ```

Agreed, that looks right. I experimented a bit with it after I wrote my last post and it didn't quite feel as awesome as I thought it would. It's not that we can't inspect, ascertain the types, and generate the stubs, but that like you mention, I think there are some devils in the details including:
1. if type stubs are generated for fixtures beside each test (or in one or more stubs directories), stubs contributed by a plugin as in the example I gave wouldn't be extensible by anything on the VS Code side, one stub will always have priority
2. if the idea was to generate e.g. `test_foo.pyi` beside an actively edited `test_foo.py`, current hinting mechanics in VS Code don't appear to apply the stub's definitions over top of the active file's own definitions
3. whichever way types are eventually hinted for fixtures, I think it would be helpful to ensure there's some "fixture" type applied to these, e.g. in your example `my_fixture` should probably have a way of being hinted as `Fixture[list[str]]` so to speak (a bit in the spirit of `Callable`)

For that second point, what I mean is if you are actively editing `test_foo.py` with code like in your example, the IntelliSense doesn't show the `test_foo.pyi` hints, it shows `my_fixture: Any` despite the stub's hint. However, if you were to put the fixture in its own module, then reference it from another file, yes, it would pick up the stub's hint. That's an unfortunate wrinkle though I'm sure it makes complete sense.

**Comment by erictraut** [comment]
Yeah, I don't see how type stubs help here — for the reasons you mention above (especially point 2).

To get intellisense for parameters in any other function, we require that developers provide type annotations for input parameters. Increasingly, Python developers are adopting this habit because they see the benefits not only for completion suggestions but also for static type checking. Pytest unit tests are simply functions, so I would think that we'd want to follow the same pattern here and require developers to add inlined type annotations for parameters.

The big difference is that we have some additional type information about the arguments that are likely to be passed to each parameter (by using type inference on the return type of the fixture or parameterization marker). That means we could offer a code action to insert the inferred type as a type annotation for the parameter, saving the developer the extra work.

This approach would maintain consistency of experience (since test functions would be like every other function), and it wouldn't violate any of the other core principles that we established with pylance and pyright. In particular, we want parity between diagnostics produced in pylance and in pyright (assuming the two are configured the same). This is critical for developers who want to automate testing & type checking, for example through the use of CI scripts.

**Comment by jwilges** [comment]
> Pytest unit tests are simply functions, so [...] we'd want to follow the same pattern here and require developers to add inlined type annotations for parameters.

The `pytest` runner is performing a form of dependency injection where it calls fixture functions based on test function matching argument names at runtime and passes the fixture return values into the test function. If the editor were to simply offer auto-completion in the form of injecting a once-off type annotation for these arguments, the inserted type annotations could easily become out of date when someone updates any associated fixture's return value type hints. That's why there was a discussion earlier about regenerating the type hints on save vs. statically injecting them into the test module source.

> The big difference is that we have some additional type information about the arguments that are likely to be passed to each parameter (by using type inference on the return type of the fixture or parameterization marker). That means we could offer a code action to insert the inferred type as a type annotation for the parameter, saving the developer the extra work.

In my inline response above I'm assuming by "code action to insert" you mean as I've described above, which has the aforementioned pitfall(s). If you mean to insert the inferred type dynamically as part of IntelliSense despite the test function not specifying a type hint, disregard my above concern. Can you clarify your idea if I misinterpreted it?

**Comment by gramster** [comment]
Linking some past related issues:

https://github.com/microsoft/pylance-release/issues/356
https://github.com/microsoft/pylance-release/issues/684
https://github.com/microsoft/pylance-release/issues/1059
https://github.com/microsoft/pylance-release/issues/1750
https://github.com/microsoft/pylance-release/issues/1824

We've also had [this issue](https://github.com/microsoft/pylance-release/issues/3614) opened about 5 times, but it's different to fixture type inference and involves code flow, so not related to 3727.

**Comment by bschnurr** [comment]
This issue has been fixed in prerelease version 2023.2.11, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023211-1-february-2023-prerelease)
